# Data Science and Machine Learning Pretest

*   This Colab is read-only. Please save a copy of it on your Drive to edit it by going to `Menu > File > Save a copy in Drive`.
*   Rename your Colab in the following format and replace 111423000 with your student ID:
> `Copy 111423000 of Data Science and Machine Learning pretest v21.02.ipynb`
*   You are required to complete this pretest **on your own**.
*   You MUST provide a list of referred resources at the end of your answers.
*   When you have completed the questions below, make sure you turn on the **share/edit/view persmission**.


# Question 1: Text preprocessing
Most of the text data acquired through web crawling and review can be noisy. When handling this kind of text data, preprocessing is an important step to ensure the quality of the dataset. There are several procedures in text preprocessing, including:
1. lowercase 
2. decontracting 
3. remove tags, punctuations, numbers
3. tokenization 
4. stopword removal 
5. lemmatization 
6. stemming


 

## 1-1. Please briefly explain what each step does and describe how you would apply them in a text classification task. Use examples/practical applications in your answers. (30%)


1.lowercase : 將英文字母改為小寫字母。

為什麼做lwercase或不做lowercase?

在大部分狀況，會在前處理時進行lowercase的動作，可幫助後續進行NLP處理上的效能[1]，也因此大部分的人都會將自己的所有資料都進行lowercase的處理。[2]lowercase經常被忽略，但是最基本最簡單的前處理方式，可以協助資料探勘與進行NLP處理時，讓輸出的結果有一致性。當在進行neural network等模型訓練時，如果訓練集過小，則會造成在進行test時因為大小寫差異導致無輸出結果或是輸出較少的情形，舉例TAIWAN、Taiwan、taiwan、TAiwan，可能被視為不同的內容。[3]

不做lowercase的原因，在辨識JAVA與python的程式語言，其撰寫規則不同，JAVA的格式與python的格式，可以使用大小寫來進行初步判斷，如此時已進行lowercase處理，則少了一項特徵。[3]

2.decontracting

因為英文的語法，有縮寫與一般寫法的不同，為了讓兩種文字為同一種意思，會將其進行decontracting，以避免在進行text classification task時，同樣意思卻視為不同語詞，舉例如"can't"等同於"can not"，"yor're"等同於"you are"。[4]


3.remove tags, punctuations, numbers

當在進行資料檢索、text classification時我們會刪去tags, punctuations, numbers，因為在進行分類時，非文字以外的數字不視為特徵，舉例在進行google搜尋時，如果搜尋"10種機器學習的方式"，則可能會出現多種的機器學習方式頁面，不限定於只有10種，可能會大於等於或小於，而下方則會顯示關鍵字缺少:10，利用，利用remove tags, punctuations, numbers，可以上，可以讓資料檢索、text classification的範圍不會被文字以外的特徵限縮。[5]

4.tokenization

是將字串進行切割，當我們要判斷的text不是字母、句子、文章而是一個單字時，則使用此方法進行切割，以不同的語言則會有不同的方式進行，舉例I love text classificatioon.是一個句子，利用tokenizaation則可將句子切割成"I"、"love"、"text"、"classification"，以進行後續NLP處理[6]

5.stopword removal

當我們利用google進行資料檢索時，通常我們會直接下關鍵字進行搜尋，或是整個句子，但其實像是冠詞、動詞等字並非重要的檢索資訊，則可以進行前處理刪除，已找到更多的資訊，舉例: what is stopword removal，我們可以簡化成 stopword removal以進行後續搜尋。通常可以應用於search systems、text classification。利用stopword removal的動作可以減少搜尋不相關的關鍵字，以提高搜尋或是分類速度。[3]

6.lemmatization

此項前處理和stemming相似，差別在於進行的是如"good"、"well"是一樣的意思，將其整合視為一種字詞，通常會使用dictionary，如: WordNet，以及一些規則進行處理，但其處理成效較stemming小，並且在進行neural network時，此處理會花費比stemming更多的時間，但影響成效不大，是否須對text進行lemmatization，可取決於使用的演算法以及時間成本是否有益。[3]

7.stemming

字詞的不同語態整合，例如：classes，class，一個為複數單位，一個為單一單位，但其實是指一樣的字詞，當我們在搜尋 "deep learning classes"，我們希望也能找到"deep learn class"等相關字詞，因此需要使用stemming來整合，以進行後續NLP處理，而進行stemming的格式有很多種，其中最常見的是Porters Algorithm[3]
Reference

[1] https://www.oreilly.com/library/view/python-natural-language/9781787121423/9742008f-6384-42a4-9711-2721dd6fd382.xhtml

[2] https://www.kaggle.com/competitions/billion-word-imputation/discussion/10862

[3] https://www.kdnuggets.com/2019/04/text-preprocessing-nlp-machine-learning.html

[4] https://basilkjose.medium.com/data-preprocessing-natural-language-competition-processing-dcbbf9d014e8

[5] https://blog.ekbana.com/pre-processing-text-in-python-ad13ea544dae

[6] https://www.analyticsvidhya.com/blog/2020/05/what-is-tokenization-nlp/

## 1-2. Please apply text preprocessing to the following sample text provided below.(70%)

In [ ]:
documents = ["Product arrived labeled as Jumbo Salted Peanuts...the peanuts were actually small sized unsalted. Not sure if this was an error or if the vendor intended to represent the product as 'Jumbo'",
        "I WAS VISITING MY FRIEND NATE THE OTHER MORNING FOR COFFEE , HE CAME OUT OF HIS STORAGE ROOM WITH ( A PACKET OF McCANNS INSTANT IRISH OATMEAL .) HE SUGGESTED THAT I TRY IT FOR MY OWN USE ,IN MY STASH . SOMETIMES NATE DOSE NOT GIVE YOU A CHANCE TO SAY NO , SO I ENDED UP TRYING THE APPLE AND CINN . FOUND IT TO BE VERY TASTEFULL WHEN MADE WITH WATER OR POWDERED MILK . IT GOES GOOD WITH O.J. AND COFFEE AND A SLICE OF TOAST AND YOUR READY TO TAKE ON THE WORLD...OR THE DAY AT LEAST..  JERRY REITH...",
        "I don't know if it's the cactus or the tequila or just the unique combination of ingredients, but the flavour of this hot sauce makes it one of a kind!  We picked up a bottle once on a trip we were on and brought it back home with us and were totally blown away!  When we realized that we simply couldn't find it anywhere in our city we were bummed.<br /><br />Now, because of the magic of the internet, we have a case of the sauce and are ecstatic because of it.<br /><br />If you love hot sauce..I mean really love hot sauce, but don't want a sauce that tastelessly burns your throat, grab a bottle of Tequila Picante Gourmet de Inclan.  Just realize that once you taste it, you will never want to use any other sauce.<br /><br />Thank you for the personal, incredible service!",
        "Product received is as advertised.<br /><br /><a href='http://www.amazon.com/gp/product/B001GVISJM'>Twizzlers, Strawberry, 16-Ounce Bags (Pack of 6)</a>",
        "this was sooooo deliscious but too bad i ate em too fast and gained 2 pds! my fault",
        "Deal was awesome!  Arrived before Halloween as indicated and was enough to satisfy trick or treaters.  I love the quality of this product and it was much less expensive than the local store's candy.",
        "I love these.........very tasty!!!!!!!!!!!  Infact, I think I am addicted to them.<br />Buying them in packs of 6 bags - is very reasonable than going to Target and getting a bag.  Savings are about a $1.00 a bag.  I use subscribe and save on these and several other product.  I love subscribe and save!!!!!!!!!!!",
        "I LOVE spicy ramen, but for whatever reasons this thing burns my stomach badly and the burning sensation doesn't go away for like 3 hours! Not sure if that is healthy or not .... and you can buy this at Walmart for $0.28, way cheaper than Amazon.",
        "Makes a tasty, super easy meal, fast. BUT high in calories.<br /><br />The instructions say to saute the veggies first but I recommend cooking the chicken first. The chicken takes longer to cook and the raw chicken ontop of veggies just makes a slimy mess. I made it with snow peas and carrots only. I dont like the little corn.  Added some red pepper flakes for heat and served ontop of rice.  It came out wonderful! Dinner on the table in less than 30mins.",
        "Love this sugar.  I also get muscavado sugar and they are both great to use in place of regular white sugar. Recommend!",
        "This is just Fantastic Chicken Noodle soup, the best I have ever eaten, with large hearty chunks of chicken,and vegetables and nice large noodles. This soup is just so full bodied, and is seasoned just right.  I am so glad Amazon carries this product.  I just can't find it here in Vermont."]

In [ ]:
import pandas as pd
import nltk
import re
nltk.download('punkt')
nltk.download('wordnet')
nltk.download('omw-1.4')
nltk.download('stopwords')
from nltk.stem import WordNetLemmatizer
from nltk.tokenize import sent_tokenize, word_tokenize
from nltk.stem import PorterStemmer
from nltk.corpus import stopwords
from bs4 import BeautifulSoup
df= pd.DataFrame(documents,columns=['sentence'])
##########3.remove tags, punctuations, numbers ########
# ref: https://michael-fuchs-python.netlify.app/2021/05/22/nlp-text-pre-processing-i-text-cleaning/#removing-irrelevant-characters-numbers-and-punctuation
def remove_tags_puncs_nums(sentence):
    out1 = BeautifulSoup(sentence, 'html.parser').get_text()
    out2 = re.sub(r'[^a-zA-Z]', ' ', sentence)
    return out2
##########5. stopword removal ########
# ref: https://www.geeksforgeeks.org/removing-stop-words-nltk-python/
def stop_words(sentence):
  stop_words = set(stopwords.words('english'))
  word_tokens = word_tokenize(sentence)
  filtered_sentence = []  
  for w in word_tokens:
    if w not in stop_words:
      filtered_sentence.append(w)
      filtered_sentence.append(" ")
  return "".join(filtered_sentence)

##########6. lemmatization ########
# ref: https://www.datacamp.com/tutorial/stemming-lemmatization-python
wordnet_lemmatizer = WordNetLemmatizer()
punctuations="?:!.,;"

def lemmaSentence(sentence):
  sentence_words = nltk.word_tokenize(sentence)
  lemma_sentence=[]
  for word in sentence_words:
      # if word in punctuations:
      #     sentence_words.remove(word)      
    lemma_sentence.append(wordnet_lemmatizer.lemmatize(word, pos="v"))
    lemma_sentence.append(" ")
  return "".join(lemma_sentence)

########7. stemming#########
# ref: https://www.datacamp.com/tutorial/stemming-lemmatization-python
porter = PorterStemmer()
porter=PorterStemmer()

def stemSentence(sentence):
    token_words = word_tokenize(sentence)
    token_words
    stem_sentence=[]
    for word in token_words:
        stem_sentence.append(porter.stem(word))
        stem_sentence.append(" ")
    return "".join(stem_sentence)

##########1. lowercase################
df['sentence'] = df['sentence'].astype(str)

for i in range(len(df)):
  df['sentence'][i] = df['sentence'][i].lower()
  ##########2. decontracting###############
  # ref: https://www.kaggle.com/code/hachemsfar/real-or-not-data-clearning-preprocessing、https://stackoverflow.com/questions/19790188/expanding-english-language-contractions-in-python
  df['sentence'][i] = df['sentence'][i][:].replace("won't", "will not").replace("can't", "can not")
  df['sentence'][i] = df['sentence'][i][:].replace("n't", " not").replace("'s", " is").replace("'t", " not").replace("'re", " are").replace("'d", " would").replace("'ll", " will").replace("'ve", " have").replace("'m", " am")
  # call 3.remove tags, punctuations, numbers 的function
  df['sentence'][i] = remove_tags_puncs_nums(df['sentence'][i])
  # call 5. stopword removal 的function
  df['sentence'][i] = stop_words(df['sentence'][i])
  # call 6. lemmatization 的function
  df['sentence'][i] = lemmaSentence(df['sentence'][i])
  # call 7. stemming 的function
  df['sentence'][i] = stemSentence(df['sentence'][i])
  ##########4. tokenization#############
  df['sentence'][i] = sent_tokenize(df['sentence'][i])
df



[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


sentence
0   [product arriv label jumbo salt peanut peanut ...
1   [visit friend nate morn coffe come storag room...
2   [know cactu tequila uniqu combin ingredi flavo...
3   [product receiv advertis br br href http www a...
4      [sooooo delisci bad eat em fast gain pd fault]
5   [deal awesom arriv halloween indic enough sati...
6   [love tasti infact think addict br buy pack ba...
7   [love spici ramen whatev reason thing burn sto...
8   [make tasti super easi meal fast high calori b...
9   [love sugar also get muscavado sugar great use...
10  [fantast chicken noodl soup best ever eat larg...

# Question 2: DataFrame handling
***For Question 2, you may NOT import any additional libraries/packages to solve the questions.***



*   Please download the datasets from the following link, https://www.kaggle.com/aaron7sun/stocknews
*   Save the downloaded files on your own drive and load the file for later use.
- There are two channels of data provided in this dataset:

  - **News data:** Crawled historical news headlines from Reddit WorldNews Channel . They are ranked by reddit users' votes, and only the top 25 headlines are considered for a single date. (Range: 2008-06-08 to 2016-07-01)

  - **Stock data:** Dow Jones Industrial Average (DJIA) is used to "prove the concept". (Range: 2008-08-08 to 2016-07-01)



In [ ]:
from google.colab import drive
drive.mount('/content/drive', force_remount=True)

Mounted at /content/drive


In [ ]:
import pandas as pd
corpus_root = 'drive/My Drive/Colab Notebooks/datasets/' 

In [ ]:
news_df = pd.read_csv(corpus_root+'RedditNews.csv')
stock_df = pd.read_csv(corpus_root+'upload_DJIA_table.csv')

In [ ]:
news_df.head()

Date                                               News
0  2016-07-01  A 117-year-old woman in Mexico City finally re...
1  2016-07-01   IMF chief backs Athens as permanent Olympic host
2  2016-07-01  The president of France says if Brexit won, so...
3  2016-07-01  British Man Who Must Give Police 24 Hours' Not...
4  2016-07-01  100+ Nobel laureates urge Greenpeace to stop o...

In [ ]:
stock_df.head(15)

Date          Open          High           Low         Close  \
0   2016-07-01  17924.240234  18002.380859  17916.910156  17949.369141   
1   2016-06-30  17712.759766  17930.609375  17711.800781  17929.990234   
2   2016-06-29  17456.019531  17704.509766  17456.019531  17694.679688   
3   2016-06-28  17190.509766  17409.720703  17190.509766  17409.720703   
4   2016-06-27  17355.210938  17355.210938  17063.080078  17140.240234   
5   2016-06-24  17946.630859  17946.630859  17356.339844  17400.750000   
6   2016-06-23  17844.109375  18011.070312  17844.109375  18011.070312   
7   2016-06-22  17832.669922  17920.160156  17770.359375  17780.830078   
8   2016-06-21  17827.330078  17877.839844  17799.800781  17829.730469   
9   2016-06-20  17736.869141  17946.359375  17736.869141  17804.869141   
10  2016-06-17  17733.439453  17733.439453  17602.779297  17675.160156   
11  2016-06-16  17602.230469  17754.910156  17471.289062  17733.099609   
12  2016-06-15  17703.650391  17762.960938  17629.009766  17640.169922   
13  2016-06-14  17710.769531  17733.919922  17595.789062  17674.820312   
14  2016-06-13  17830.500000  17893.279297  17731.349609  17732.480469   

       Volume     Adj Close  
0    82160000  17949.369141  
1   133030000  17929.990234  
2   106380000  17694.679688  
3   112190000  17409.720703  
4   138740000  17140.240234  
5   239000000  17400.750000  
6    98070000  18011.070312  
7    89440000  17780.830078  
8    85130000  17829.730469  
9    99380000  17804.869141  
10  248680000  17675.160156  
11   91950000  17733.099609  
12   94130000  17640.169922  
13   93740000  17674.820312  
14  101690000  17732.480469


## 2-1. Define a *funtion* aims to know the weekly (7 days) Stock **Close value** trend. (80%)
- Implement a *funtion* that determines the weekly (7 days) Stock **Close value** trend using the Stock data and **then record the trend in the News data in a new column "label" with 0, 1 and -1**. For example, 
  - On 2016-06-13, the market closed at 17732.480469. Seven days later, 2016-06-20, the market closed **higher** at 17804.869141. In this scenario, all entries corresponding to 2016-06-13 in the News data will be **marked 1** in the "label" column. Dates 2016-06-14 and 2016-06-15 will also be marked 1 because 2016-06-21 and 2016-06-22 closed higher, respectively.
  - On the other hand, 2016-06-17 will be **marked 0** because 2016-06-24 (7 days later) closed **lower**.
  - If a given date does not have a corresponding date for 7 days later, the given date will be **marked -1**.




In [ ]:
import datetime
def label(first_day_close, sevenDays_after_close):
  '''
  please answer here, you can add any parameters if you want.
  but don't import other libraries, this notebook already prepared the libraries which all you need !
  remember that, the standard for evaluation include your:
    1. Time Complexity (80%)
    2. Program Logic (10%)
    3. Creativity (10%)
  '''
  if sevenDays_after_close + first_day_close == -2:
    return -1
  elif sevenDays_after_close > first_day_close:
    return 1
  else :
    return 0



In [ ]:
'''
than apply your defined function on stock_df here.
return results store into a new columns name "Label"

'''
def stock_df_count(stock_df):
  sum_record = len(stock_df)
  stock_df['Label']=""
  for i in range(sum_record-5):
    # ref: https://blog.51cto.com/u_12461028/5575525
    # 將字串轉成時間格式date
    first_day_column = sum_record - (i+1) 
    seven_day_column = first_day_column - 5
    first_day = datetime.datetime.strptime(stock_df['Date'][first_day_column], "%Y-%m-%d").date()
    for i in range(5):
      seven_day = datetime.datetime.strptime(stock_df['Date'][seven_day_column+i], "%Y-%m-%d").date()
      days = (seven_day - first_day).days     # 計算天數差
      if days == 7:
        stock_df.loc[:,('Label')][first_day_column] = label(stock_df['Close'][first_day_column], stock_df['Close'][seven_day_column])
        break
      elif stock_df['Label'][first_day_column] != (0 or -1) :
        stock_df.loc[:,('Label')][first_day_column] = label(-1, -1)
  for i in range(5):
    stock_df.loc[:,('Label')][i] = label (-1, -1)
    
  return stock_df['Label']
stock_df_count(stock_df)


<ipython-input-9-85fe1b29865b>:19: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  stock_df.loc[:,('Label')][first_day_column] = label(stock_df['Close'][first_day_column], stock_df['Close'][seven_day_column])
<ipython-input-9-85fe1b29865b>:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  stock_df.loc[:,('Label')][first_day_column] = label(-1, -1)
<ipython-input-9-85fe1b29865b>:24: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  stock_df.l

0       -1
1       -1
2       -1
3       -1
4       -1
        ..
1984     0
1985     0
1986     0
1987     0
1988     0
Name: Label, Length: 1989, dtype: object

## 2-2. Map your label to news data (20%)

In [ ]:
def news_label(news_df, stock_df):
  '''
  Map your close value trend label in news data by date, also name it "Label"
  if a date does not appear in the stock data, also label it as -1.
  e.g. all news in 2016-06-30 will be label -1
  '''
  len(news_df)
  news_df.groupby(["Date"]).size()
  news_df['Label']= "-1"
  the_day = 0
  one_day_news = 25
  the_label_day = 0
  while(the_day < (len(news_df)) and the_label_day < (len(stock_df))):
    stock_day = datetime.datetime.strptime(stock_df['Date'][the_label_day], "%Y-%m-%d").date()
    news_day = datetime.datetime.strptime(news_df['Date'][the_day], "%Y-%m-%d").date()
    if stock_day == news_day:
      news_df['Label'][the_day:the_day+25] = stock_df['Label'][the_label_day]
      the_day = the_day + one_day_news
      the_label_day = the_label_day + 1
    elif (stock_day - news_day).days < 0: # 計算天數差
      news_df['Label'][the_day:the_day+25] = -1
      the_day = the_day + one_day_news

    else: 
      the_label_day = the_label_day + 1
  return news_df['Label']
news_label(news_df, stock_df)



0        -1
1        -1
2        -1
3        -1
4        -1
         ..
73603    -1
73604    -1
73605    -1
73606    -1
73607    -1
Name: Label, Length: 73608, dtype: object

# Question 3: Compute cosine similarity of TF-IDF (term frequency–inverse document frequency)
***For Question 3, you may NOT import any additional libraries/packages to solve the questions.***
-  **Cosine similarity** is a measure of similarity between two non-zero vectors of an inner product space that measures the cosine of the angle between them. 
- **TF-IDF** is a numerical statistic that is intended to reflect how important a word is to a document in a collection or corpus








## 3-1. Cosine similarity and Euclidean distance are two commonly used metrics to quantify the similarity between two vectors in a vector space model. Explain the differences and implications when we interpret the similarity scores measured by these two metrics. Use examples in your answers. (20%)

Cosine similarity、Euclidean distance相異處與範例:

Cosine similarity

計算兩向量之間的相似度，藉由計算兩向量的餘弦夾角度數來進行分析。 在Document similarity進行文檔的相似度探討、Recommender systems根據用戶的喜好推薦其書籍或電影，將每個用戶的評分、價格等視為向量屬性進行分析、Image processing分析由不同角度拍攝同一物品的兩張照片、Clustering、Text classification等處理中都會使用。

Euclidean distance

主要用於計算兩點之間的直線距離。 在進行Image processing中，可以將圖像的像素轉為矩陣，後進行距離計算，以分析他們的差異度、、Dimensionality reduction中希望將高維數據以低維的方式呈現並保留其原始特性，則會以Euclidean distance計算其差異性，以進行後續映射的動作、Signal processing中以時間的序列差異來評估其差異性，而其他如Recommender systems、Clustering也可以利用此方法來進行分析。

兩者雖都是在計算兩向量或是兩點之間的差異性、相似度，但因為其公式的特性，Cosine similarity較適用於進行text classification、recommender systems以及 clustering。
Euclidean distance則較適用於image processing, signal processing, and dimensionality reduction。因公式計算差異，Cosine similarity的值介於-1-1之間，而Euclidean distance的值則為0-∞，

舉例以recommender systems來說，兩者雖會有相似的差異性分類推薦，但其表示的量表差度數字則有大小程度之分，所以其對進行推薦系統的有利程度亦不同。其他不同的應用也如此範例所述

Reference

[1]https://www.sciencedirect.com/topics/computer-science/cosine-similarity

[2]https://www.cuemath.com/euclidean-distance-formula/


## 3-2. Define a converting function to compute tf-idf vector from a list of ducoments. (40%)

In [ ]:
documents = ['terrible service this time','terrible terrible service','most terrible service','terrible service and experience','what a terrible service','so terrible service experience','what a terrible disappointment','what a terrible place','this time it was so horrible','the staff was horrible']

In [ ]:
'''
Answer here
you can define other functions to support the defined function if you need.

TF-IDF dataframe show as the following table.
'''
import numpy as np
import math
def get_tf(doc):
    tf_dictionary = {}
    for word in doc:
        if word in tf_dictionary:
            tf_dictionary[word] += 1
        else:
            tf_dictionary[word] = 1
    for word in tf_dictionary:
        tf_dictionary[word] = tf_dictionary[word] / float(len(doc))
    return tf_dictionary

def get_idf(documents):
  idf_dict = {}
  word_lists = [doc.split() for doc in documents]
  unique_words = set(word for word_list in word_lists for word in word_list)
  words = np.concatenate(word_lists)
  unique_words = np.unique(words)
  for the_term_index in range(len(unique_words)):
    term = unique_words[the_term_index]
    n = sum(1 for doc in documents if term in doc.split()) ####doc.split()非常重要，將會變成找字母或是找單字
    N = len(documents)
    idf_dict[term] = math.log(N / n)  
  return idf_dict

def computTFIDF(documents):
  tfidf = []
  idf = get_idf(documents)
  for doc in documents:
      tf = get_tf(doc.split())
      tfidf_doc = {'terrible':float(0), 'service':float(0), 'this':float(0),'time':float(0), 'most':float(0),'and':float(0),'experience':float(0),'what':float(0),'a':float(0)
      ,'so':float(0),'disappointment':float(0),'place':float(0),'it':float(0),'was':float(0),'horrible':float(0),'the':float(0),'staff':float(0)}
      for term, val in tf.items():
          tfidf_doc[term] = val * idf[term]
      tfidf.append(tfidf_doc)

  return np.array(tfidf).tolist()

In [ ]:
import pandas as pd
tf_idf_list = computTFIDF(documents)
df = pd.DataFrame(computTFIDF(documents))
df

terrible   service      this      time      most       and  experience  \
0  0.055786  0.127706  0.402359  0.402359  0.000000  0.000000    0.000000   
1  0.148762  0.170275  0.000000  0.000000  0.000000  0.000000    0.000000   
2  0.074381  0.170275  0.000000  0.000000  0.767528  0.000000    0.000000   
3  0.055786  0.127706  0.000000  0.000000  0.000000  0.575646    0.402359   
4  0.055786  0.127706  0.000000  0.000000  0.000000  0.000000    0.000000   
5  0.055786  0.127706  0.000000  0.000000  0.000000  0.000000    0.402359   
6  0.055786  0.000000  0.000000  0.000000  0.000000  0.000000    0.000000   
7  0.055786  0.000000  0.000000  0.000000  0.000000  0.000000    0.000000   
8  0.000000  0.000000  0.268240  0.268240  0.000000  0.000000    0.000000   
9  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000    0.000000   

       what         a        so  disappointment     place        it       was  \
0  0.000000  0.000000  0.000000        0.000000  0.000000  0.000000  0.000000   
1  0.000000  0.000000  0.000000        0.000000  0.000000  0.000000  0.000000   
2  0.000000  0.000000  0.000000        0.000000  0.000000  0.000000  0.000000   
3  0.000000  0.000000  0.000000        0.000000  0.000000  0.000000  0.000000   
4  0.300993  0.300993  0.000000        0.000000  0.000000  0.000000  0.000000   
5  0.000000  0.000000  0.402359        0.000000  0.000000  0.000000  0.000000   
6  0.300993  0.300993  0.000000        0.575646  0.000000  0.000000  0.000000   
7  0.300993  0.300993  0.000000        0.000000  0.575646  0.000000  0.000000   
8  0.000000  0.000000  0.268240        0.000000  0.000000  0.383764  0.268240   
9  0.000000  0.000000  0.000000        0.000000  0.000000  0.000000  0.402359   

   horrible       the     staff  
0  0.000000  0.000000  0.000000  
1  0.000000  0.000000  0.000000  
2  0.000000  0.000000  0.000000  
3  0.000000  0.000000  0.000000  
4  0.000000  0.000000  0.000000  
5  0.000000  0.000000  0.000000  
6  0.000000  0.000000  0.000000  
7  0.000000  0.000000  0.000000  
8  0.268240  0.000000  0.000000  
9  0.402359  0.575646  0.575646

## 3-3. Define a scoring function to compute the cosine similarity between two input vector. (30%)

In [ ]:
'''
Answer here
return the cosine similarity between the given two vectors
Apply the function which you designed to all sentences, and show your scoring results as the following table.
'''
def cosine_sim(vec_a, vec_b):
  #ref: https://stackoverflow.com/questions/39608421/showing-valueerror-shapes-1-3-and-1-3-not-aligned-3-dim-1-1-dim-0
  dot_product = float(np.dot(vec_a, vec_b.T))
  mag_a = float(math.sqrt(np.dot(vec_a, vec_a.T)))
  mag_b = float(math.sqrt(np.dot(vec_b, vec_b.T)))
  score = float(dot_product / (mag_a * mag_b))
  return score

## 3-4. Show the cross comparation table for the given sentences. (10%)

In [ ]:
def cosine_similarity_of_TF_IDF(documents):
  df = pd.DataFrame(computTFIDF(documents))
  cosine_similarity_array = []
  for row in range(len(documents)):
    geted_cosine_sim = []
    for cloumn in range(len(documents)):
      geted_cosine_sim.append(cosine_sim(df[row:row+1][0:], df[cloumn:cloumn+1][0:]))
    cosine_similarity_array.append(geted_cosine_sim)
  return cosine_similarity_array

cosine_similarity_of_TF_IDF_list = cosine_similarity_of_TF_IDF(documents)
cos_sim_tf_idf_df = pd.DataFrame(cosine_similarity_of_TF_IDF(documents))
cos_sim_tf_idf_df

0         1         2         3         4         5         6  \
0  1.000000  0.226813  0.055972  0.046299  0.074014  0.056587  0.007397   
1  0.226813  1.000000  0.224349  0.185576  0.296664  0.226813  0.051111   
2  0.055972  0.224349  1.000000  0.045796  0.073209  0.055972  0.007317   
3  0.046299  0.185576  0.045796  1.000000  0.060557  0.432244  0.006053   
4  0.074014  0.296664  0.073209  0.060557  1.000000  0.074014  0.573020   
5  0.056587  0.226813  0.055972  0.432244  0.074014  1.000000  0.007397   
6  0.007397  0.051111  0.007317  0.006053  0.573020  0.007397  1.000000   
7  0.007397  0.051111  0.007317  0.006053  0.573020  0.007397  0.357407   
8  0.517451  0.000000  0.000000  0.000000  0.000000  0.258725  0.000000   
9  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   

          7         8         9  
0  0.007397  0.517451  0.000000  
1  0.051111  0.000000  0.000000  
2  0.007317  0.000000  0.000000  
3  0.006053  0.000000  0.000000  
4  0.573020  0.000000  0.000000  
5  0.007397  0.258725  0.000000  
6  0.357407  0.000000  0.000000  
7  1.000000  0.000000  0.000000  
8  0.000000  1.000000  0.305206  
9  0.000000  0.305206  1.000000